includes/class-sensei-utils.php

```diff
@@ -1598,19 +1618,32 @@ public static function user_completed_lesson( $lesson = 0, $user_id = 0 ): bool
 				if ( 0 >= (int) $user_id ) {
 					return false;
 				}
-				$_user_lesson_status = self::user_lesson_status( $lesson, $user_id );
-
-				if ( isset( $_user_lesson_status->comment_approved ) ) {
 
-					$user_lesson_status = $_user_lesson_status->comment_approved;
+				$lesson_id = (int) $lesson;
+				$user_id   = (int) $user_id;
 
+				$lesson_progress = Sensei()->lesson_progress_repository->get( $lesson_id, $user_id );
+				if ( $lesson_progress ) {
+					$user_lesson_status = $lesson_progress->get_status();
 				} else {
-
-					return false; // No status means not complete
-
+					return false; // No progress means not complete
 				}
 
-				$lesson_id = $lesson;
+				// In the comments-based progress we use one entry to store both the lesson progress and the quiz progress.
+				// In the tables-based progress we split them. Here is important to use the quiz proress if the quiz pass is required.
+				$lesson_quiz_id = Sensei()->lesson->lesson_quizzes( $lesson_id );
+				if ( $lesson_quiz_id ) {
+					$quiz_progress = Sensei()->quiz_progress_repository->get( $lesson_quiz_id, $user_id );
+					$pass_required = get_post_meta( $lesson_quiz_id, '_pass_required', true );
+
+					if ( $pass_required ) {
+						if ( $quiz_progress ) {
```

Comment:
Just a small optimization suggestion. Fetching the progress can be done after the `$pass_required` check. Potentially preventing a database call.

```suggestion
					$pass_required = get_post_meta( $lesson_quiz_id, '_pass_required', true );

					if ( $pass_required ) {
						$quiz_progress = Sensei()->quiz_progress_repository->get( $lesson_quiz_id, $user_id );
						if ( $quiz_progress ) {
```

Reply:
Maybe you've missed this feedback because Github decided to hide the conversations?

Reply:
There are a few other hidden ones as well... You need to press the "load more..." button to see them.

![image](https://github.com/Automattic/sensei/assets/1612178/dafba902-9e5e-465f-a247-6fc959f8ff14)